allauth/socialaccount/providers/apple/client.py

```diff
@@ -0,0 +1,100 @@
+import requests
+from time import time
+from urllib.parse import parse_qsl
+
+from django.conf import settings
+from django.utils.http import urlencode
+
+import jwt
+
+from allauth.socialaccount.providers.oauth2.client import (
+    OAuth2Client,
+    OAuth2Error,
+)
+
+
+class Scope(object):
+    EMAIL = 'email'
+    NAME = 'name'
+
+
+class AppleOAuth2Client(OAuth2Client):
+    """
+    Custom client because `Sign In With Apple`:
+        * requires `response_mode` field in redirect_url
+        * requires special `client_secret` as JWT
+    """
+
+    def generate_client_secret(self):
+        """Create a JWT signed with an apple provided private key"""
+        APPLE_PROVIDER_SETTINGS = getattr(
+            settings, 'SOCIALACCOUNT_PROVIDERS', {}
+        ).get('apple', {})
+
+        MEMBER_ID = APPLE_PROVIDER_SETTINGS.get('MEMBER_ID', None)
+        SECRET_KEY = APPLE_PROVIDER_SETTINGS.get('SECRET_KEY', None)
+
+        CURRENT_TIMESTAMP = int(time())
+        claims = {
+            'iss': MEMBER_ID,
+            'aud': 'https://appleid.apple.com',
+            'sub': self.consumer_key,
+            'iat': CURRENT_TIMESTAMP,
+            'exp': CURRENT_TIMESTAMP + 15777000,
+        }
+        headers = {'kid': self.consumer_secret, 'alg': 'ES256'}
+        client_secret = jwt.encode(
+            payload=claims,
+            key=SECRET_KEY,
+            algorithm='ES256',
+            headers=headers
+        ).decode('utf-8')
+        return client_secret
+
+    @classmethod
+    def is_json_response(cls, response):
+        return (
+            response.headers['content-type'].split(
+                ';')[0] == 'application/json'
+            or response.text[:2] == '{"'
+        )
+
+    def get_access_token(self, code):
+        url = self.access_token_url
+        client_secret = self.generate_client_secret()
+        data = {
+            'client_id': self.consumer_key,
+            'code': code,
+            'grant_type': 'authorization_code',
+            'redirect_uri': self.callback_url,
+            'client_secret': client_secret
+        }
+        self._strip_empty_keys(data)
+        resp = requests.request(
+            self.access_token_method,
+            url,
+            data=data,
+            headers=self.headers)
+        access_token = None
+        if resp.status_code in [200, 201]:
+            if self.is_json_response(resp):
```

Comment:
Is this helper method really needed? Couldn't you just invoke resp.json() directly and catch the ValueError in case it is not proper JSON?

Reply:
@pennersr updated by this [commit](https://github.com/pennersr/django-allauth/pull/2424/commits/25d094299f426d892a04fe521c7d186fe5510745)